:book: Docs for record events. (#3375)
kubernetes-sigs/kubebuilder: docs/book/src/reference/creating-events.md

# Creating Events

It is often useful to publish *Event* objects from the controller Reconcile function. 

Events allow users to see what is going on with a particular object, and allow automated processes to see and respond to them.

>  Recent Events for an object can be viewed by running `kubectl describe`

Reviewer: ```suggestion
 Recent Events for an object can be viewed by running `kubectl describe <resource kind> <resource name>`
```

# Creating Events

It is often useful to publish *Event* objects from the controller Reconcile function. 

Events allow users to see what is going on with a particular object, and allow automated processes to see and respond to them.

>  Recent Events for an object can be viewed by running `kubectl describe <resource kind> <resource name>`.

Reviewer: ```suggestion
 Recent Events for an object can be viewed by running `$ kubectl describe <resource kind> <resource name>`. Also, they can be checked by running `$ kubectl get events`. 
```

>  

Reviewer: ```suggestion

 <aside class="warning">
 
Events should be raised in certain circumstances only. Be aware that it is NOT recommended to emit events for all Operations. If authors raise too many events, it brings bad UX experiences for those consuming the solutions on the cluster, and they will not have attention to what they must be notified of. For further information, see the Kubernetes APIs convention describes when using them [here][Events]. 

</aside>
```

It is a very important information. Therefore, we should:
- a) add on the top to ensure that readers will read this one
- b) use the warning class because we are warning users

>   ### how to implement the solution to raise an event:

Reviewer: ```suggestion
## How to raise events? 

Following are the steps with examples to help you raise events in your controller's reconciliations. 
```

Reviewer: Let's add the definition with an event within its usage example.
And then, afterwords, provide a detailed example of how to implement it fully. 

It seems more natural. You will first read what it is, how to use it, and its purpose. You only go to the details when/if you need to.

> a) You need to pass the recorder when you set up the recorder for the controller when the event will be raised see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/main.go#L103

Reviewer: ```suggestion
### Passing the Recorder to the Controller

To raise 

```

Reviewer: Lets first change the controller; otherwise, users will check the compiler failing. 
We can also do a generic example and put the imports.

> b) You need to add the makers to add the RBAC permissions to allow your Operator/controller to raise the event see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L66 and run make manifests

Reviewer: ````suggestion
### Passing the EventRecorder to the Controller

Now, we will pass the EventRecorder when we are setting up the controller on the Manager. See that we will change the implementation scaffolded in `cmd/main.go`:

```go
	if err = (&controller.MyKindReconciler{
		Client:   mgr.GetClient(),
		Scheme:   mgr.GetScheme(),
		// Note that we added the following line: 
		Recorder: mgr.GetEventRecorderFor("mykind-controller"),
	}).SetupWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "MyKind")
		os.Exit(1)
	}
```

````

Reviewer: Now, that we changed the controller, then we can describe how to pass the struct to it.

> c) You can check an example of the event being called in: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303

Reviewer: ````suggestion
### Granting the required permissions

You must also grant the RBAC rules permissions to allow your project to create Events. Therefore, ensure that you add the [RBAC][rbac-markers] into your controller:

```go
...
//+kubebuilder:rbac:groups=core,resources=events,verbs=create;patch
...

func (r *MyKindReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

And then, run `$ make manifests` to update the rules under `config/rbac/rule.yaml`. 
````

Reviewer: We will need to add at the bottom an alias `[rbac-markers]` for the page: https://book.kubebuilder.io/reference/markers/rbac.html

It is a kubebuilder doc, so we should NOT use the above URL; instead, we should use a relative path from this place to the file.

> [rbac-markers]: markers/rbac.md

> 

Reviewer: You can remove the whole content afterwords. We added all info above already.

> Events are published from a Controller using an [EventRecorder]`type CorrelatorOptions struct`,
> which can be created for a Controller by calling `GetRecorder(name string)` on a Manager.
> 
> `Name` should be identifiable and descriptive as it will appear in the `From` column of `kubectl describe` command.
> 
>  
> ```go
> // Annotation for generating RBAC role for writing Events
> // +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
> ```
> 
> ```go
> // ReconcileContainerSet reconciles a ContainerSet object
> type ReconcileContainerSet struct {
>   client.Client
>   scheme *runtime.Scheme
>   recorder record.EventRecorder
> }
> ```

Reviewer: ```suggestion
```
We do not need anymore this text.

> ```go
> // newReconciler returns a new reconcile.Reconciler
> func newReconciler(mgr manager.Manager) reconcile.Reconciler {
>   return &ReconcileContainerSet{
>     Client:   mgr.GetClient(),
>     scheme:   mgr.GetScheme(),
>     recorder: mgr.GetRecorder("containerset-controller"),
>   }
> }
> ```
>  
> 
>  

Reviewer: ```suggestion
```
We should either remove this part

> ## Writing Events
> 
> Anatomy of an Event:
> 
> ```go
> Event(object runtime.Object, eventtype, reason, message string)
> ```
> 
> - `object` is the object this event is about.
> - `eventtype` is the type of this event, and is either *Normal* or *Warning*.
> - `reason` is the reason this event is generated. It should be short and unique with `UpperCamelCase` format. The value could appear in *switch* statements by automation.
> - `message` is intended to be consumed by humans.

Reviewer: ```suggestion
```
Lets add it on the top.

> 
> Building on the example introduced in [Controller Example](../basics/simple_controller.md), we can add Events to our reconcile logic using `recorder` as our `EventRecorder`
> 
>  
> ```go
>   //Reconcile logic up here...
> 
>   // Create the resource
>   found := &appsv1.Deployment{}
>   err = r.Get(context.TODO(), types.NamespacedName{Name: deploy.Name, Namespace: deploy.Namespace}, found)
>   if err != nil && errors.IsNotFound(err) {
>     log.Printf("Creating Deployment %s/%s\n", deploy.Namespace, deploy.Name)
>     err = r.Create(context.TODO(), deploy)
>     if err != nil {
>       return reconcile.Result{}, err
>     }
>     
>     // Write an event to the ContainerSet instance with the namespace and name of the 
>     // created deployment
>     r.recorder.Event(instance, "Normal", "Created", fmt.Sprintf("Created deployment %s/%s", deploy.Namespace, deploy.Name))
>   
>   } else if err != nil {
>     return reconcile.Result{}, err
>   }
> 
>   // Preform update
>   if !reflect.DeepEqual(deploy.Spec, found.Spec) {
>     found.Spec = deploy.Spec
>     log.Printf("Updating Deployment %s/%s\n", deploy.Namespace, deploy.Name)
>     err = r.Update(context.TODO(), found)
>     if err != nil {
>       return reconcile.Result{}, err
>     }
> 
>     // Write an event to the ContainerSet instance with the namespace and name of the 
>     // updated deployment
>     r.recorder.Event(instance, "Normal", "Updated", fmt.Sprintf("Updated deployment %s/%s", deploy.Namespace, deploy.Name))
>   
>   }
>   return reconcile.Result{}, nil
> }
> ```
> Events should be raised in certain circumstances only and following this guideline acctually is not a good practice. See what the Kubernetes APIs convention describes when using them [here][Events]
> 
> <aside class="note">
>        It is NOT recommended to emit events for all Operations. If authors raise too many events it brings bad UX experiences for those that are consuming the        solutions on the cluster and they will not have attention to what they actually must be notified.

Reviewer: ```suggestion
```

We can remove all added here.

> Building on the example introduced in [Controller Example](../basics/simple_controller.md), we can add Events to our reconcile logic using `recorder` as our `EventRecorder`
> 
>  
> ```go
>   //Reconcile logic up here...
> 
>   // Create the resource
>   found := &appsv1.Deployment{}
>   err = r.Get(context.TODO(), types.NamespacedName{Name: deploy.Name, Namespace: deploy.Namespace}, found)
>   if err != nil && errors.IsNotFound(err) {
>     log.Printf("Creating Deployment %s/%s\n", deploy.Namespace, deploy.Name)
>     err = r.Create(context.TODO(), deploy)
>     if err != nil {
>       return reconcile.Result{}, err
>     }
>     
>     // Write an event to the ContainerSet instance with the namespace and name of the 
>     // created deployment
>     r.recorder.Event(instance, "Normal", "Created", fmt.Sprintf("Created deployment %s/%s", deploy.Namespace, deploy.Name))
>   
>   } else if err != nil {
>     return reconcile.Result{}, err
>   }
> 
>   // Preform update
>   if !reflect.DeepEqual(deploy.Spec, found.Spec) {
>     found.Spec = deploy.Spec
>     log.Printf("Updating Deployment %s/%s\n", deploy.Namespace, deploy.Name)
>     err = r.Update(context.TODO(), found)
>     if err != nil {
>       return reconcile.Result{}, err
>     }
> 
>     // Write an event to the ContainerSet instance with the namespace and name of the 
>     // updated deployment
>     r.recorder.Event(instance, "Normal", "Updated", fmt.Sprintf("Updated deployment %s/%s", deploy.Namespace, deploy.Name))
>   
>   }
>   return reconcile.Result{}, nil
> }
> ```
> Events should be raised in certain circumstances only and following this guideline acctually is not a good practice. See what the Kubernetes APIs convention describes when using them [here][Events]
> 
> <aside class="note">
>        It is NOT recommended to emit events for all Operations. If authors raise too many events it brings bad UX experiences for those that are consuming the        solutions on the cluster and they will not have attention to what they actually must be notified.
> </aside>

Reviewer: ```suggestion
```

# Creating Events

It is often useful to publish *Event* objects from the controller Reconcile function. 

Events allow users to see what is going on with a particular object, and allow automated processes to see and respond to them.

 Recent Events for an object can be viewed by running `$ kubectl describe <resource kind> <resource name>`. Also, they can be checked by running `$ kubectl get events`. 

<aside class="warning">
<h1>Events should be raised in certain circumstances only</h1>

> Be aware that it is **not** recommended to emit Events for all Operations. If authors raise too many events, it brings bad UX experiences for those consuming the solutions on the cluster, and they will not have attention to what they must be notified of. For more information, please take a look at the [Kubernetes APIs convention][Events]. 

Reviewer 2: ```suggestion
Be aware that it is **not** recommended to emit Events for all operations. If authors raise too many events, it brings bad UX experiences for those consuming the solutions on the cluster, and they may find it difficult to filter an actionable event from the clutter. For more information, please take a look at the [Kubernetes APIs convention][Events]. 
```

</aside>

## Writing Events

Anatomy of an Event:

```go
Event(object runtime.Object, eventtype, reason, message string)
```

- `object` is the object this event is about.
- `eventtype` is this event type, and is either *Normal* or *Warning*.
- `reason` is the reason this event is generated. It should be short and unique with `UpperCamelCase` format. The value could appear in *switch* statements by automation.

> - `message` is intended to be consumed by humans.

Reviewer 2: Can we add links for each of the type specifications so that its easier. For example:
event type would be https://github.com/kubernetes/api/blob/6c11c9e4685cc62e4ddc8d4aaa824c46150c9148/core/v1/types.go#L6019-L6024
reason would be https://github.com/kubernetes/api/blob/6c11c9e4685cc62e4ddc8d4aaa824c46150c9148/core/v1/types.go#L6048 
message would be https://github.com/kubernetes/api/blob/6c11c9e4685cc62e4ddc8d4aaa824c46150c9148/core/v1/types.go#L6053

Reviewer: @Sajiyah-Salat just add the links in the above statements like; 

[(More info)][link-alias] after the explanation I think that would be enough.

Otherwise, all is great for me here and this one has my lgtm

GREAT WORK

Reviewer: Can you please squash all commits?
git rebase -i HEAD~<number of commits>
Then, you replace pick with s to squash the commit with the previous one.

	r.Recorder.Event(cr, "Warning", "Deleting",
		fmt.Sprintf("Custom Resource %s is being deleted from the namespace %s",
			cr.Name,
			cr.Namespace))

```

</aside>

## How to be able to raise Events?

Following are the steps with examples to help you raise events in your controller's reconciliations. 

a) You need to pass the recorder when you set up the recorder for the controller when the event will be raised see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/main.go#L103
### Passing the EventRecorder to the Controller

Events are published from a Controller using an [EventRecorder]`type CorrelatorOptions struct`,
which can be created for a Controller by calling `GetRecorder(name string)` on a Manager. See that we will change the implementation scaffolded in `cmd/main.go`:

```go
	if err = (&controller.MyKindReconciler{
		Client:   mgr.GetClient(),
		Scheme:   mgr.GetScheme(),
		// Note that we added the following line: 
		Recorder: mgr.GetEventRecorderFor("mykind-controller"),
	}).SetupWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "MyKind")
		os.Exit(1)
	}
```

> c) You can check an example of the event being called in: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303

Reviewer 2: Nit:
```suggestion
c) You can check an example of the event being called in [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303) 
```

> c) You can check an example of the event being called in: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303

Reviewer 2: Nit:
```suggestion
c) You can check an example of the event being called in [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/controllers/memcached_controller.go#L299-L303) 
```

# Creating Events

It is often useful to publish *Event* objects from the controller Reconcile function as they allow users or any automated processes to see what is going on with a particular object and respond to them.

 Recent Events for an object can be viewed by running `$ kubectl describe <resource kind> <resource name>`. Also, they can be checked by running `$ kubectl get events`. 

<aside class="warning">
<h1>Events should be raised in certain circumstances only</h1>

Be aware that it is **not** recommended to emit Events for all operations. If authors raise too many events, it brings bad UX experiences for those consuming the solutions on the cluster, and they may find it difficult to filter an actionable event from the clutter. For more information, please take a look at the [Kubernetes APIs convention][Events]. 

</aside>

## Writing Events

Anatomy of an Event:

```go
Event(object runtime.Object, eventtype, reason, message string)
```

- `object` is the object this event is about.

> - `eventtype` is this event type, and is either *Normal* or *Warning*. [More info][Event-Example]

Reviewer: ```suggestion
- `eventtype` is this event type, and is either *Normal* or *Warning*. ([More info][Event-Example])
```

> - `reason` is the reason this event is generated. It should be short and unique with `UpperCamelCase` format. The value could appear in *switch* statements by automation. [More info][[Reason-Example]]

Reviewer: ```suggestion
- `reason` is the reason this event is generated. It should be short and unique with `UpperCamelCase` format. The value could appear in *switch* statements by automation. ([More info][[Reason-Example]])
```

> - `message` is intended to be consumed by humans. [More info][Message-Example]

Reviewer: ```suggestion
- `message` is intended to be consumed by humans. ([More info][Message-Example])
```

<aside class="note">
<h1>Example Usage</h1>

Following is an example of a code implementation that raises an Event.

```go
	// The following implementation will raise an event
	r.Recorder.Event(cr, "Warning", "Deleting",
		fmt.Sprintf("Custom Resource %s is being deleted from the namespace %s",
			cr.Name,
			cr.Namespace))

```

</aside>

> 

Reviewer: @Sajiyah-Salat 

It is part of the explanation over how to achieve this goal using the scaffold provided by the Kubebuilder as it is missing parts provided in the suggestions.
The idea is to provide the 3 steps that users must to do for they know how to do that with Kubebuilder projects.

Reviewer: See the previous suggestions: https://github.com/kubernetes-sigs/kubebuilder/pull/3375#discussion_r1187045581

> Events are published from a Controller using an [EventRecorder]`type CorrelatorOptions struct`,

Reviewer: @Sajiyah-Salat now see that you would replace the a, b, c for the steps.
It means here we would have the first step as suggested: https://github.com/kubernetes-sigs/kubebuilder/pull/3375#discussion_r1187046166

- `eventtype` is this event type, and is either *Normal* or *Warning*.
- `reason` is the reason this event is generated. It should be short and unique with `UpperCamelCase` format. The value could appear in *switch* statements by automation.
- `message` is intended to be consumed by humans.

> More [info](https://github.com/kubernetes/api/blob/6c11c9e4685cc62e4ddc8d4aaa824c46150c9148/core/v1/types.go#L6019-L6024):

Reviewer: @Sajiyah-Salat see that each spec in the Kubernertes Api Event is described and point out by varsha in the previous suggestion comment.

Please, ensure that you click on in each link to check them out and add the [[more info][link]] for each case scenario.

Also, remember that ALL links should be added as alias and the alias should defined at the bottom.

By last, do not forget to squash all commits and that you addressed all comments and suggestions
